spyder_kernels/console/shell.py

```diff
@@ -157,48 +163,117 @@ def remove_pdb_session(self, pdb_obj):
 
+    def add_namespace_manager(self, ns_manager):
+        """Add namespace manager to stack."""
+        self._namespace_stack.append(ns_manager)
+
+    def remove_namespace_manager(self, ns_manager):
+        """Remove namespace manager."""
+        if self._namespace_stack[-1] != ns_manager:
+            logger.debug("The namespace stack is inconsistent.")
+            return
+        self._namespace_stack.pop()
+
+    def get_local_scope(self, stack_depth):
+        """
+        Get local scope at given frame depth.
+
+        Needed for magics that use "needs_local_scope" such as timeit
+        """
+        frame = sys._getframe(stack_depth + 1)
+        return self.context_locals(frame)
+
+    def context_locals(self, frame=None):
+        """
+        Get context locals.
+
+        If frame is not None, make sure frame.f_locals is not registered in a
+        debugger and return frame.f_locals
+        """
+        for session in self._namespace_stack[::-1]:
+            if isinstance(session, SpyderPdb) and session.curframe is not None:
+                if frame is None or frame == session.curframe:
+                    return session.curframe_locals
+            elif frame is None and isinstance(session, NamespaceManager):
+                return session.ns_locals
+        if frame is not None:
+            return frame.f_locals
+        return None
+
     @property
     def _pdb_frame(self):
         """Return current Pdb frame if there is any"""
         if self.pdb_session is not None:
             return self.pdb_session.curframe
 
-    @property
-    def _pdb_locals(self):
-        """
-        Return current Pdb frame locals if available. Otherwise
-        return an empty dictionary
-        """
-        if self._pdb_frame is not None:
-            return self.pdb_session.curframe_locals
-        else:
-            return {}
-
     @property
     def user_ns(self):
         """Get the current namespace."""
-        if self._pdb_frame is not None:
-            return self._pdb_frame.f_globals
-        else:
-            return self.__user_ns
+        for session in self._namespace_stack[::-1]:
+            if isinstance(session, SpyderPdb) and session.curframe is not None:
+                # return first debugging namespace
+                return session.curframe.f_globals
+            elif isinstance(session, NamespaceManager):
+                return session.ns_globals
+        return self.__user_ns
 
     @user_ns.setter
     def user_ns(self, namespace):
         """Set user_ns."""
         self.__user_ns = namespace
 
+    def _get_current_namespace(self, with_magics=False, frame=None):
+        """
+        Return a copy of the current namespace
+
+        Only return a single dict for variable exxplorer.
+        This is globals() if not debugging, or a dictionary containing
+        both locals() and globals() for current frame when debugging
+        """
```

Comment:
```suggestion
        """Return a copy of the current namespace."""
```

The rest of this docstring doesn't make sense anymore.